### PR TITLE
Pydanticgen: support polymorphism using type designators

### PR DESCRIPTION
**Pull request by sierra-moxon**
This is a PR that echoes https://github.com/linkml/linkml/pull/1180, 1180 was on a development branch and was reviewed/accepted by Kevin already, this is "to main" version.

### CODE REVIEW

File: linkml/generators/common/__init__.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from linkml_runtime.linkml_model.meta import SlotDefinition, ClassDefinition
+from linkml_runtime.utils.schemaview import SchemaView
+from linkml_runtime.utils.formatutils import camelcase
+
+def get_type_designator_value(sv: SchemaView, type_designator_slot: SlotDefinition, class_def: ClassDefinition) -> str:
```
Comment (cmungall) [inline linkml/generators/common/__init__.py:5]: is the __init__.py the best place for this?

File: tests/test_issues/test_pydantic_polymorphism.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        assert len(output_subset) > 0
+        assert "const=True" in output_subset[0]
+        assert len([x for x in output_subset if 'http://example.org/Person' in x]) == 1
+
+    def test_pydantic_load_poly_data(self):
+        gen = PydanticGenerator(schema_str)
+        output = gen.serialize()
+        mod = compile_python(output, "testschema")
+        data = mod.Container.parse_raw(data_str)
+        assert len([x for x in data.things if type(x) == mod.Person]) == 1
```
Comment (cmungall) [inline tests/test_issues/test_pydantic_polymorphism.py:100]: minor stylistic point: I prefer `self.assert<SpecificMethod>`, and also explicit isinstance checks

**Review by cmungall — APPROVED** [review]
This is fantastic, thanks!!

File: linkml/generators/common/__init__.py (hunk 1 above)
Comment (kervel) [reply linkml/generators/common/__init__.py:5]: shall i put it in a separate `type_designators.py` under `linkml/generators/common` ? can i edit it directly in this branch ?

Comment (kevinschaper) [reply linkml/generators/common/__init__.py:5]: `linkml/generators/common/type_designators.py` sounds like a good place - I think you should be able to edit the branch, let us know if you can't.

**Comment by kervel** [comment]
resolved all comments ...

**Comment by kevinschaper** [comment]
It can be a bit of a challenge to find what part of the test output actually shows the failure:

but it looks like it's the new pydantic inlining test
```
FAIL: test_pydantic_inlining (tests.test_generators.test_pydanticgen.PydanticGeneratorTestCase)
[78949](https://github.com/linkml/linkml/actions/runs/3933217266/jobs/6726684206#step:9:78950)
----------------------------------------------------------------------
[78950](https://github.com/linkml/linkml/actions/runs/3933217266/jobs/6726684206#step:9:78951)
Traceback (most recent call last):
[78951](https://github.com/linkml/linkml/actions/runs/3933217266/jobs/6726684206#step:9:78952)
  File "D:\a\linkml\linkml\tests\test_generators\test_pydanticgen.py", line 214, in test_pydantic_inlining
[78952](https://github.com/linkml/linkml/actions/runs/3933217266/jobs/6726684206#step:9:78953)
    self.assertIn(f"a2b: {expected}", slot_line)
[78953](https://github.com/linkml/linkml/actions/runs/3933217266/jobs/6726684206#step:9:78954)
AssertionError: 'a2b: Optional[List[str]]' not found in '    a2b: Optional[Dict[str, str]] = Field(default_factory=list)'
```

**Comment by kervel** [comment]
i rebased on main, seems this was missing a bugfix that was done already in main.

**Pull request merged**